src/aatoolbox/utils/raster.py

```diff
@@ -0,0 +1,324 @@
+"""Utilities to manipulate and analyze raster data."""
+
+import logging
+from typing import Any, Dict, List, Union
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import rioxarray
+import xarray as xr
+
+logger = logging.getLogger(__name__)
+
+
+def invert_coordinates(
+    ds: Union[xr.Dataset, xr.DataArray],
+    lon_coord: str,
+    lat_coord: str,
+):
+    """
+    Invert latitude and longitude in ``ds``.
+
+    This function checks for inversion of latitude and longitude
+    and inverts them if needed. Datasets with inverted coordinates
+    can produce incorrect results in certain functions like
+    ``rasterstats.zonal_stats()``. Correctly ordered coordinates
+    should be:
+
+    * latitude: Largest to smallest.
+    * longitude: Smallest to largest.
+
+    If ``ds`` already has correct coordinate ordering, it is
+    directly returned. Function largely copied from
+    https://github.com/perrygeo/python-rasterstats/issues/218.
+
+    Parameters
+    ----------
+        ds : Union[xarray.DataArray, xarray.Dataset]
+            Dataset with values and coordinates.
+        lon_coord : str
+            Longitude coordinate in ``ds``.
+        lat_coord : str
+            Latitude coordinate in ``ds``.
+
+    Returns
+    -------
+    Union[xarray.DataArray, xarray.Dataset]
+        Dataset or data array with correct coordinate ordering.
+
+    Examples
+    --------
+    >>> da = xarray.DataArray(
+         numpy.arange(16).reshape(4,4),
+         coords={"lat":numpy.array([87, 88, 89, 90]),
+                 "lon":numpy.array([70, 69, 68, 67])}
+    )
+    >>> invert_coordinates(ds, "lon", "lat")
+    """
+    lon_inv, lat_inv = _check_coords_inverted(
+        ds=ds, lon_coord=lon_coord, lat_coord=lat_coord
+    )
+
+    # Flip the raster as necessary (based on the flags)
+    inv_dict = {}
+
+    if lon_inv:
+        logger.info("Longitude was inverted, reversing coordinates.")
+        inv_dict[lon_coord] = ds[lon_coord][::-1]
+
+    if lat_inv:
+        logger.info("Latitude was inverted, reversing coordinates.")
+        inv_dict[lat_coord] = ds[lat_coord][::-1]
+
+    if inv_dict:
+        ds = ds.reindex(inv_dict)
+
+    return ds
+
+
+def _check_coords_inverted(ds, lon_coord, lat_coord):
+    """Check if latitude and longitude inverted."""
+    lat_start = ds[lat_coord][0].item()
+    lat_end = ds[lat_coord][ds.dims[lat_coord] - 1].item()
+    lon_start = ds[lon_coord][0].item()
+    lon_end = ds[lon_coord][ds.dims[lon_coord] - 1].item()
+    return lon_start > lon_end, lat_start < lat_end
+
+
+# TODO: understand when to change longitude range for rasterstats
+# to work and when not!!!
```

Comment:
Haha I like exclamation marks in my old comments. But think this can be removed now. As far as I understand rasterstats overlays the array, and incorporates the CRS but this isn't able to account for an incorrect range. This is one of the main reasons we want to rely more on the rioxarray functions for computing stats than rasterstats. At the same time, if we know the range is not in common conventions, we prefer to change it just to be sure

Reply:
In the end, I have removed the doctest implementation until we can sort out #23.